### PR TITLE
topologymanager: Promote support for improved multi-numa alignment in Topology Manager to beta

### PR DESCRIPTION
**Pull request by PiotrProkop**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:

Promote `prefer-closest-numa-nodes` policy option to Beta and enable by default.

In order to do so, we have to make sure that `TopologyManagerPolicyOptions` and `TopologyManagerPolicyBetaOptions` feature gates are enabled by default to be able to serve this capability.

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
TopologyManagerPolicyOptions feature-flag is promoted to beta and enabled by default.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

- KEP:  https://github.com/kubernetes/enhancements/pull/4079
- Enhancement issue: https://github.com/kubernetes/enhancements/issues/3545

```docs

```

### CODE REVIEW

**Comment by PiotrProkop** [comment]
/retest pull-kubernetes-verify

seems like a timeout

```
There are no nodes that your pod can schedule to - check your requests, tolerations, and node selectors (0/51 nodes are available: 1 node(s) had untolerated taint {node.kubernetes.io/not-ready: }, 50 Insufficient cpu, 8 Insufficient memory. preemption: 0/51 nodes are available: 1 Preemption is not helpful for scheduling, 50 No preemption victims found for incoming pod.)
```

**Comment by ffromani** [comment]
/triage accepted
/prioprity important-soon

priority because this work is part of the 1.28 cycle

File: pkg/features/kube_features.go [hunk 1]
```diff
@@ -1052,9 +1052,9 @@ var defaultKubernetesFeatureGates = map[featuregate.Feature]featuregate.FeatureS
 
 	TopologyManagerPolicyAlphaOptions: {Default: false, PreRelease: featuregate.Alpha},
 
-	TopologyManagerPolicyBetaOptions: {Default: false, PreRelease: featuregate.Beta},
+	TopologyManagerPolicyBetaOptions: {Default: true, PreRelease: featuregate.Beta},
```
Comment (ffromani) [inline pkg/features/kube_features.go:1055]: AFAIU this is a "bug fix" meaning that this should have defaulted to `true` since the beginning. If so I'd isolate the change in its own commit

File: pkg/kubelet/cm/topologymanager/policy_options_test.go [hunk 2]
```diff
@@ -28,7 +28,7 @@ import (
 	pkgfeatures "k8s.io/kubernetes/pkg/features"
 )
 
-var fancyBetaOption = "fancy-new-option"
+var fancyAlphaOption = "fancy-new-option"
```
Comment (ffromani) [inline pkg/kubelet/cm/topologymanager/policy_options_test.go:31]: why demote from beta to alpha?

File: pkg/kubelet/cm/topologymanager/policy_options_test.go [hunk 3]
```diff
@@ -39,55 +39,67 @@ type optionAvailTest struct {
...
 		},
 		{
-			description: "test beta options success",
+			description: "test alpha options fail",
 			policyOptions: map[string]string{
-				fancyBetaOption: "true",
+				fancyAlphaOption: "true",
 			},
-			expectedErr: fmt.Errorf("Topology Manager Policy Beta-level Options not enabled,"),
+			expectedErr: fmt.Errorf("Topology Manager Policy Alpha-level Options not enabled,"),
```
Comment (ffromani) [inline pkg/kubelet/cm/topologymanager/policy_options_test.go:93]: this looks OK, but do we want to keep the old test? (not replace it)

File: pkg/kubelet/cm/topologymanager/topology_manager.go [hunk 4]
```diff
@@ -133,13 +133,13 @@ var _ Manager = &manager{}
...
 func NewManager(topology []cadvisorapi.Node, topologyPolicyName string, topologyScopeName string, topologyPolicyOptions map[string]string) (Manager, error) {
-	klog.InfoS("Creating topology manager with policy per scope", "topologyPolicyName", topologyPolicyName, "topologyScopeName", topologyScopeName)
-
 	opts, err := NewPolicyOptions(topologyPolicyOptions)
 	if err != nil {
 		return nil, err
 	}
 
+	klog.InfoS("Creating topology manager with policy per scope", "topologyPolicyName", topologyPolicyName, "topologyScopeName", topologyScopeName, "topologyPolicyOptions", opts)
+
```
Comment (ffromani) [inline pkg/kubelet/cm/topologymanager/topology_manager.go:142]: why did we move the log?

**Review by ffromani — COMMENTED** [review]
looks reasonnable, mostly questions inside to make sure I understood all the details

File: pkg/features/kube_features.go (hunk 1 above)
Comment (PiotrProkop) [reply pkg/features/kube_features.go:1055]: Yeah, good idea :)

File: pkg/kubelet/cm/topologymanager/policy_options_test.go (hunk 2 above)
Comment (PiotrProkop) [reply pkg/kubelet/cm/topologymanager/policy_options_test.go:31]: Cause we don't have any `alpha` features when we are graduating `prefer-closest-numa` to beta and I need at least one `alpha` feature to test out corner cases.

File: pkg/kubelet/cm/topologymanager/policy_options_test.go (hunk 3 above)
Comment (PiotrProkop) [reply pkg/kubelet/cm/topologymanager/policy_options_test.go:93]: We have previous test but it had to be changed a bit to accomodate that not beta feature gate is enabled:
```go
	{
			description: "fail to set option when TopologyManagerPolicyBetaOptions feature gate is not set",
			featureGate: pkgfeatures.TopologyManagerPolicyBetaOptions,
			policyOptions: map[string]string{
				PreferClosestNUMANodes: "true",
			},
			expectedErr: fmt.Errorf("Topology Manager Policy Beta-level Options not enabled,"),
		},
```

File: pkg/kubelet/cm/topologymanager/topology_manager.go (hunk 4 above)
Comment (PiotrProkop) [reply pkg/kubelet/cm/topologymanager/topology_manager.go:142]: I wanted to print out `topologyPolicyOptions`, I can also leave old log and add another message when parsing opts is successful, WDYT?

Comment (ffromani) [reply pkg/kubelet/cm/topologymanager/topology_manager.go:142]: I think this is a good enough reasons for the move, we can keep it (provided there are no side effects. I think there are not, let's just doublecheck for peace of mind)

File: pkg/features/kube_features.go (hunk 1 above)
Comment (PiotrProkop) [reply pkg/features/kube_features.go:1055]: This change was moved to separate commit.

File: pkg/kubelet/cm/topologymanager/topology_manager.go (hunk 4 above)
Comment (PiotrProkop) [reply pkg/kubelet/cm/topologymanager/topology_manager.go:142]: The only change is that kubelet won't log policy and scope if `NewPolicyOptions` is failing.
Previously it would log them and fail afterwards.

File: pkg/features/kube_features.go (hunk 1 above)
Comment (ffromani) [reply pkg/features/kube_features.go:1055]: thanks, this part LGTM

File: pkg/features/kube_features.go [hunk 5]
```diff
@@ -1054,7 +1054,7 @@ var defaultKubernetesFeatureGates = map[featuregate.Feature]featuregate.FeatureS
 
 	TopologyManagerPolicyBetaOptions: {Default: true, PreRelease: featuregate.Beta},
 
-	TopologyManagerPolicyOptions: {Default: false, PreRelease: featuregate.Alpha},
+	TopologyManagerPolicyOptions: {Default: true, PreRelease: featuregate.Beta},
```
Comment (ffromani) [inline pkg/features/kube_features.go:1057]: please highlight in the commit message and/or in the PR description that we want 
- to promote "topology manager policy options exist at all" (or any better phrasing you can think of) from alpha to beta. The reason I'm highlighting this is we had some initial confusion when doing the same work for cpumanager, "*ManagerPolicyOptions" and "*ManagerPolicyOptions{Alpha,Beta}" easily creates confusion
- at the same time, promote the "PreferClosestNUMANodes" option from alpha to beta

I don't think you should issue separate PRs, but I do think that clear phrasing in the commit messages/PR description and perhaps more granular commit split will make the intent clearer.

File: pkg/kubelet/cm/topologymanager/policy_options_test.go (hunk 2 above)
Comment (ffromani) [reply pkg/kubelet/cm/topologymanager/policy_options_test.go:31]: ok, can we add (vs replace) a `fancyBetaOption` to be able to increase the test coverage?

File: pkg/kubelet/cm/topologymanager/topology_manager.go (hunk 4 above)
Comment (ffromani) [reply pkg/kubelet/cm/topologymanager/topology_manager.go:142]: right, but the error is bubbled up. I think we can keep the change you made and move on.

File: pkg/kubelet/cm/topologymanager/topology_manager_test.go [hunk 6]
```diff
@@ -86,10 +87,59 @@ func TestNewManager(t *testing.T) {
 				"unknown-option": "true",
 			},
 		},
+		{
```
Comment (ffromani) [inline pkg/kubelet/cm/topologymanager/topology_manager_test.go:90]: this (welcome) change doesn't seem to be stricly needed for the feature graduation, thus should be moved in its own commit

**Review by ffromani — COMMENTED** [review]
I think we're there. Comments inside about repacking the existing work for better clarity, once that is answered/addressed I think LGTM.

File: pkg/kubelet/cm/topologymanager/topology_manager_test.go (hunk 6 above)
Comment (PiotrProkop) [reply pkg/kubelet/cm/topologymanager/topology_manager_test.go:90]: You're right. I will move this to separate commit something along the line `Increase TopologyManager test coverage by adding negative test cases around NUMA topology discovery`

File: pkg/kubelet/cm/topologymanager/policy_options_test.go (hunk 2 above)
Comment (PiotrProkop) [reply pkg/kubelet/cm/topologymanager/policy_options_test.go]: I have changed the code according to your review.

File: pkg/kubelet/cm/topologymanager/policy_options_test.go (hunk 3 above)
Comment (PiotrProkop) [reply pkg/kubelet/cm/topologymanager/policy_options_test.go]: added old test

File: pkg/kubelet/cm/topologymanager/topology_manager.go (hunk 4 above)
Comment (PiotrProkop) [reply pkg/kubelet/cm/topologymanager/topology_manager.go:142]: 👍

File: pkg/kubelet/cm/topologymanager/topology_manager_test.go (hunk 6 above)
Comment (PiotrProkop) [reply pkg/kubelet/cm/topologymanager/topology_manager_test.go:90]: I have moved this change into seperate commit.

**Review by ffromani — COMMENTED** [review]
/lgtm

thanks @PiotrProkop for updating and addressing all my comments

**Comment by ffromani** [comment]
/assign @klueska 

for final review and approval

**Review by swatisehgal — COMMENTED** [review]
Please update the PR description to indicate that the end goal is to have `prefer-closest-numa-nodes` policy option promoted to Beta and enabled by default. In order to do so, we HAVE to ensure that the overall feature and the feature gates guarding its enablement ( namely`TopologyManagerPolicyOptions`  and `TopologyManagerPolicyBetaOptions`)   are enabled by default to serve this capability .

Perhaps `Promote support for improved multi-numa alignment in Topology Manager to beta` would be a better title of this PR?

Please note that the`Experimental` prefix in its definition in  container manager should be dropped in this PR too:

https://github.com/kubernetes/kubernetes/blob/2f563464bc879cc7d31642eb77e2b0d38b498159/pkg/kubelet/cm/container_manager.go#L160-L161

**Comment by PiotrProkop** [comment]
> Please update the PR description to indicate that the end goal is to have `prefer-closest-numa-nodes` policy option promoted to Beta and enabled by default. In order to do so, we HAVE to ensure that the overall feature and the feature gates guarding its enablement ( namely`TopologyManagerPolicyOptions` and `TopologyManagerPolicyBetaOptions`) are enabled by default to serve this capability .
> 
> Perhaps `Promote support for improved multi-numa alignment in Topology Manager to beta` would be a better title of this PR?
> 
> Please note that the`Experimental` prefix in its definition in container manager should be dropped in this PR too:
> 
> https://github.com/kubernetes/kubernetes/blob/2f563464bc879cc7d31642eb77e2b0d38b498159/pkg/kubelet/cm/container_manager.go#L160-L161

Will do. Thanks for comments!

**Review by swatisehgal — APPROVED** [review]
/lgtm

**Comment by swatisehgal** [comment]
/assign @dims
for features approval

**Comment by PiotrProkop** [comment]
@klueska I had to rebase due to conflicts.

**Comment by PiotrProkop** [comment]
Failures seems unrelated, unit tests in `pkg/kubelet/cm/topologymanager` are green. `pull-kubernetes-node-e2e-containerd` complains about timeouts.

/retest

**Comment by ffromani** [comment]
/assign @dchen1107 

for feature approval and final review

**Comment by dims** [comment]
/approve

(pkg/features!)
